cdc/log.cc

```diff
@@ -1258,6 +1258,52 @@ class transformer final : public change_processor {
     // Takes and returns generated cdc log mutations and associated statistics about parts touched during transformer's lifetime.
     // The `transformer` object on which this method was called on should not be used anymore.
     std::tuple<std::vector<mutation>, stats::part_type_set> finish() && {
+        static const auto preimg_op_bytes  = _op_col.type->decompose(operation_native_type(operation::pre_image));
+        static const auto postimg_op_bytes = _op_col.type->decompose(operation_native_type(operation::post_image));
+
+        // Adjust or remove deltas according to `delta_mode` in `cdc_options`.
+        if (_schema->cdc_options().get_delta_mode() != cdc::delta_mode::full) {
+            for (auto& m : _result_mutations) {
+                auto& clustered_rows = m.partition().clustered_rows();
+                int deleted_rows_cnt = 0;
+                for (auto it = clustered_rows.begin(); it != clustered_rows.end(); /* no increment */) {
+                    const auto& op_cell = it->row().cells().cell_at(_op_col.id).as_atomic_cell(_op_col);
+                    const auto op_val = op_cell.value().linearize();
+                    if (op_val != preimg_op_bytes && op_val != postimg_op_bytes) {
+                        if (_schema->cdc_options().get_delta_mode() == cdc::delta_mode::off) {
+                            it = m.partition().clustered_rows().erase_and_dispose(it, current_deleter<rows_entry>());
+                            ++deleted_rows_cnt;
+                            continue;
+                        }
+
+                        // The case of `get_delta_mode() == delta_mode::keys`:
+                        it->row().cells().remove_if([this, log_s = m.schema()] (column_id id, atomic_cell_or_collection& acoc) {
+                            const auto log_cdef = log_s->column_at(column_kind::regular_column, id);
+                            if (is_cdc_metacolumn_name(log_cdef.name_as_text())) {
+                                return false; // Keep "cdc$*" columns in the delta row.
+                            }
+                            const auto* base_cdef = _schema->get_column_definition(log_cdef.name());
+                            // Remove columns from delta that correspond to non-PK/CK columns in the base table.
+                            return base_cdef != nullptr
+                                    && (base_cdef->kind != column_kind::partition_key && base_cdef->kind != column_kind::clustering_key);
```

Comment:
Nitpick: Could we have a static_column here? If the answer is no (I'm not sure...), then it's a tiny bit faster to check if kind != regular column, than to check kind==partition_key || kind==clustering_key.

Reply:
We could